README.md
# CurrencyLayerApi

Currency Layer API is a light wrapper for [currencylayer.com](https://currencylayer.com/) for Rubyists.

## Installation

Install the gem and add to the application's Gemfile by executing:

    $ bundle add currency_layer_api

If bundler is not being used to manage dependencies, install the gem by executing:

    $ gem install currency_layer_api

## Usage

### API Documentation

The API documentation can be found [here](https://currencylayer.com/documentation).

### Get Account Information

Create an account on [currencylayer.com](https://currencylayer.com/) and get your API access key.
Save the access key in an environment variable named `CURRENCY_LAYER_ACCESS_KEY`.

### Configure

Currency Layer does not allow you ssl endpoint unless you have a paid account. By default ssl is enabled. To disable ssl, set the `ssl` option to `false`.

```ruby
require 'currency_layer_api'
client = CurrencyLayerApi::Client.new(access_key: ENV['CURRENCY_LAYER_ACCESS_KEY'], ssl: true)
```
params:
- `access_key` - Required - String - Your API access key.
- `ssl` - Optional - Boolean - Enable or disable ssl. Default is `true`.

response:
```bash
#<CurrencyLayerApi::Client:0x00007f9b0a0b0a00 @access_key="YOUR_ACCESS_KEY", @ssl=true>
```

### Get List of currencies

```ruby
client.list
```

response:

```bash
[
  #<CurrencyLayerApi::Currency:0x00007f9b0a0b0a00 @code="AED", @name="United Arab Emirates Dirham">,
  #<CurrencyLayerApi::Currency:0x00007f9b0a0b09c0 @code="AFN", @name="Afghan Afghani">,
  ...
]
```

### Get Live Exchange Rates

```ruby
client.live(from: "USD", to: ["AED", "EUR"])
```
params:
- `from` Required - String - The currency you want to convert from. (default: USD)
- `to`  Optional - Array[String] - The currencies you want to convert to. (default: nil)

response:

```bash
[
  #<CurrencyLayerApi::Conversion:0x00007f9b0a0b0a00 @from_code="USD", @to_code="AED", @value=3.6725, @date="2019-01-01">,
  #<CurrencyLayerApi::Conversion:0x00007f9b0a0b09c0 @from_code="USD", @to_code="EUR", @value=0.888, @date="2019-01-01">,
]
```

### Get Historical Exchange Rates

```ruby
client.historical(from: "USD", to: ["AED", "EUR"], date: "2019-01-01")
```
params:
- `from` Required - String - The currency you want to convert from. (default: USD)
- `to`  Optional - Array[String] - The currencies you want to convert to. (default: nil)
- `date` Required - String 'YYYY-MM-DD' - The date you want to get the exchange rates for. (default: nil)

response:

```bash
[
  #<CurrencyLayerApi::Conversion:0x00007f9b0a0b0a00 @from_code="USD", @to_code="AED", @value=3.6725, @date="2019-01-01">,
  #<CurrencyLayerApi::Conversion:0x00007f9b0a0b09c0 @from_code="USD", @to_code="EUR", @value=0.888, @date="2019-01-01">,
]
```

### Get Conversion Rate given amount

```ruby
client.convert(from: "USD", to: "AED", amount: 100)
```

params:
- `from` Required - String - The currency you want to convert from. (default: USD)
- `to`  Required - String - The currency you want to convert to. (default: nil)
- `amount` Required - Int - The amount you want to convert. (default: nil)
- `date` Optional - String 'YYYY-MM-DD' - The date you want to get the exchange rates for. (default: nil)

response:

```bash
367.25
```

### Error Handling

```ruby
begin
  client.live(from: "USD", to: ["AED", "EUR"])
rescue CurrencyLayerApi::Error => e
    puts e.code
    puts e.message
end
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/currency_layer_api. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/[USERNAME]/currency_layer_api/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the CurrencyLayerApi project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/currency_layer_api/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [0.1.1] - 2022-09-01

- Resolve changelog
- Added support for `/list` endpoint
- Added support for `/live` endpoint
- Added support for `/historical` endpoint
- Added support for `/convert` endpoint

## [0.1.0] - 2022-09-01

- First Version with basic functionality


lib/currency_layer_api/client.rb
# frozen_string_literal: true

# rubocop:disable Layout/LineLength
module CurrencyLayerApi
  # Client having API methods and access_keys
  class Client
    BASE_URL = "api.currencylayer.com/"

    attr_reader :access_key, :adapter

    # #### params:
    # access_key: String (Required)
    # ssl: Boolean (Optional) (Default: true)
    # adapter: Symbol (Optional) (Default: :net_http)
    # stubs: Faraday::Adapter::Test::Stubs (Optional) (Default: nil)
    # #### Example:
    # CurrencyLayerApi::Client.new(access_key: "fake", ssl: false, adapter: :test, stubs: stub)
    # CurrencyLayerApi::Client.new(access_key: "fake")
    # #### Description:
    # Creates a new instance of CurrencyLayerApi::Client with access_key provided by currencylayer.com
    def initialize(access_key:, ssl: true, adapter: Faraday.default_adapter, stubs: nil)
      @access_key = access_key
      @adapter = adapter
      @ssl = ssl

      # Test stubs for requests
      @stubs = stubs
    end

    # #### Example:
    # CurrencyLayerApi::Client.new(access_key: "fake").list
    # #### Response:
    # [
    #   #<CurrencyLayerApi::Currency:0x00007f9b0a0b0a00 @code="AED", @name="United Arab Emirates Dirham">,
    #   #<CurrencyLayerApi::Currency:0x00007f9b0a0b09c0 @code="AFN", @name="Afghan Afghani">,
    # ]
    # #### Description:
    # Resturns currency object array for all currencies available in currency layer
    def list
      get("list").body["currencies"].map { |key, val| Currency.new(code: key, name: val) }
    end

    # #### Params:
    # from: String (Required) (Default: "USD")
    # to: Array (Optional) (Default: nil)
    # #### Example:
    # CurrencyLayerApi::Client.new(access_key: "fake").live(from: "USD", to: ["AED", "EUR"])
    # CurrencyLayerApi::Client.new(access_key: "fake").live(from: "EUR")
    # #### Response:
    # [
    #   #<CurrencyLayerApi::Conversion:0x00007f9b0a0b0a00 @from_code="USD", @to_code="AED", @value=3.6725, @date="2019-01-01">,
    #   #<CurrencyLayerApi::Conversion:0x00007f9b0a0b09c0 @from_code="USD", @to_code="EUR", @value=0.888, @date="2019-01-01">,
    # ]
    # #### Description:
    # Returns the current exchange rate for a given from currency and to currencies
    def live(from: "USD", to: nil)
      get("live", { source: from, currencies: to ? to.join(",") : nil }.compact).body["quotes"].map do |key, val|
        Conversion.new(from_code: key[0..2], to_code: key[3..], value: val, date: Date.today.to_s)
      end
    end

    # #### Params:
    # date: String (Required) ('YYYY-MM-DD')
    # from: String (Required) (Default: "USD")
    # to: Array (Optional) (Default: nil)
    # #### Example:
    # CurrencyLayerApi::Client.new(access_key: "fake").historical(date: "2019-01-01", from: "USD", to: ["AED", "EUR"])
    # CurrencyLayerApi::Client.new(access_key: "fake").historical(date: "2019-01-01", from: "EUR")
    # #### Response:
    # [
    #   #<CurrencyLayerApi::Conversion:0x00007f9b0a0b0a00 @from_code="USD", @to_code="AED", @value=3.6725, @date="2019-01-01">,
    #   #<CurrencyLayerApi::Conversion:0x00007f9b0a0b09c0 @from_code="USD", @to_code="EUR", @value=0.888, @date="2019-01-01">,
    # ]
    # #### Description:
    # Returns the historical exchange rate for a given date, from currency and to currencies
    def historical(date:, from: "USD", to: nil)
      resp = get("historical", { date: date, source: from, currencies: to ? to.join(",") : nil }.compact)
      resp.body["quotes"].map do |key, val|
        Conversion.new(from_code: key[0..2], to_code: key[3..], value: val, date: resp.body["date"])
      end
    end

    # #### Params:
    # to: String (Required) (Currency code)
    # from: String (Required) (Currency code)
    # amount: Int (Required) (Default: 1)
    # date: String (Optional) ('YYYY-MM-DD')
    # #### Example:
    # CurrencyLayerApi::Client.new(access_key: "fake").convert(to: "AED", from: "USD", amount: 1)
    # CurrencyLayerApi::Client.new(access_key: "fake").convert(to: "AED", from: "USD", amount: 1, date: "2019-01-01")
    # #### Response:
    # 3.6725
    # #### Description:
    # Returns the exchange rate for a given date, from currency and to currency
    def convert(to:, from:, amount: 1, date: nil)
      get("convert", { from: from, to: to, amount: amount, date: date }.compact).body["result"]
    end

    private

    # creates a faraday connection with base url and access_key
    def connection
      @connection ||= Faraday.new("#{@ssl ? "https" : "http"}://#{BASE_URL}") do |conn|
        conn.params["access_key"] = access_key
        conn.request :json

        conn.response :json, content_type: "application/json"

        conn.adapter adapter, @stubs
      end
    end

    # makes a get request to the given path with params
    def get(path, params = {})
      handle_response connection.get(path, params)
    end

    # handles the response from the api and raises error if any
    # returns the response if no error
    # Error codes: https://currencylayer.com/documentation
    # error.code: 101, error.message: "You have not supplied an API Access Key. [Required format: access_key=YOUR_ACCESS_KEY]"
    def handle_response(response)
      return response if response.body["success"] == true

      raise Error.new(response.body["error"]["info"], response.body["error"]["code"])
    end
  end
end
# rubocop:enable Layout/LineLength
